enclave_build/src/pull.rs

```diff
@@ -0,0 +1,171 @@
+    EnclaveBuildError, Result,
+};
+
+/// Number of accepted image layer media types.
+pub const ACCEPTED_MEDIA_TYPES_COUNT: usize = 3;
+/// Accepted image layer media types.
+pub const ACCEPTED_MEDIA_TYPES: [&str; ACCEPTED_MEDIA_TYPES_COUNT] = [
+    oci_distribution::manifest::WASM_LAYER_MEDIA_TYPE,
+    oci_distribution::manifest::IMAGE_DOCKER_LAYER_GZIP_MEDIA_TYPE,
+    oci_distribution::manifest::IMAGE_LAYER_GZIP_MEDIA_TYPE,
+];
+/// Sequence of environment variables and config file paths to try parsing: (env, path)
+pub const CONFIG_ENV_PATHS: [(&str, &str); 3] = [
+    ("DOCKER_CONFIG", "/"),
+    ("XDG_RUNTIME_DIR", "/containers/auth.json"),
+    ("HOME", "/.docker/config.json"),
+];
+
+/// Builds a client which uses the protocol given as parameter.
+///
+/// Client required for the https://github.com/krustlet/oci-distribution library API.
+///
+/// By default, the client pulls the image matching the current running architecture.
+pub fn build_client(protocol: ClientProtocol) -> Client {
+    let client_config = ClientConfig {
+        protocol,
+        ..Default::default()
+    };
+    Client::new(client_config)
+}
+
+/// Checks the DOCKER_CONFIG env variable for the path to the docker config file
+///
+/// If DOCKER_CONFIG is not set, try to get the config file from the default path
+/// {XDG_RUNTIME_DIR}/containers/auth.json or {HOME}/.docker/config.json in a Podman-like behavior
+fn get_docker_config_file() -> Result<File> {
+    for (env, path) in CONFIG_ENV_PATHS {
+        if let Ok(env_val) = std::env::var(env) {
+            let config_path = Path::new(&env_val).join(path);
+            match File::open(&config_path) {
+                Ok(file) => {
+                    return Ok(file);
+                }
+                Err(e) => {
+                    eprintln!("Could not get credentials from {:?}: {}", config_path, e);
+                    continue;
+                }
+            }
+        }
+    }
+
+    Err(EnclaveBuildError::FileError(std::io::Error::new(
+        std::io::ErrorKind::NotFound,
+        "Config file not present",
+    )))
+}
+
+/// Returns the Docker credentials by reading from the Docker config.json file.
+///
+/// The assumed format of the file is:\
+/// {\
+///        "auths": {\
+///            "https://index.docker.io/v1/": {\
+///                    "auth": "<token_string>"\
+///            }\
+///        }\
+/// }
+pub fn parse_credentials() -> Result<RegistryAuth> {
+    let config_file = get_docker_config_file()?;
+
+    let config_json: serde_json::Value = deserialize_from_reader(&config_file).map_err(|err| {
+        EnclaveBuildError::CredentialsError(format!("JSON was not well-formatted: {:?}", err))
+    })?;
+
+    let auths = config_json.get("auths").ok_or_else(|| {
+        EnclaveBuildError::CredentialsError("Could not find auths key in config JSON".to_string())
+    })?;
+
+    if let serde_json::Value::Object(auths) = auths {
+        for registry_auths in auths.values() {
```

Comment:
Previously in `docker.rs` we checked the registry name against a registry in a specified reference.
How will it work now? Shouldn't it be the same?

Reply:
This assumes that the only credential present is for the supplied URI. I guess it should check like in `docker.rs`